src/mongocxx/private/search_index_view.hh

```diff
@@ -0,0 +1,192 @@
+#include <mongocxx/private/client_session.hh>
+#include <mongocxx/private/libbson.hh>
+#include <mongocxx/private/libmongoc.hh>
+
+#include <mongocxx/config/private/prelude.hh>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+class search_index_view::impl {
+   public:
+    impl(mongoc_collection_t* collection, mongoc_client_t* client)
+        : _coll{collection}, _client{client} {}
+
+    impl(const impl& i) = default;
+
+    impl(impl&& i) = default;
+
+    ~impl() = default;
+
+    impl& operator=(const impl& i) = default;
+
+    cursor list(const client_session* session,
+                bsoncxx::string::view_or_value name,
+                const options::aggregate& options) {
+        pipeline pipeline{};
+        pipeline.append_stage(
+            make_document(kvp("$listSearchIndexes", make_document(kvp("name", name.view())))));
+        return list(session, pipeline, options);
+    }
+
+    cursor list(const client_session* session, const options::aggregate& options) {
+        pipeline pipeline{};
+        pipeline.append_stage(make_document(kvp("$listSearchIndexes", make_document())));
+        return list(session, pipeline, options);
+    }
+
+    cursor list(const client_session* session,
+                const pipeline& pipeline,
+                const options::aggregate& options) {
+        bsoncxx::builder::basic::document opts_doc;
+        libbson::scoped_bson_t stages(bsoncxx::document::view(pipeline.view_array()));
+        const mongoc_read_prefs_t* rp_ptr = NULL;
+
+        append_aggregate_options(options, opts_doc);
+
+        if (session) {
+            opts_doc.append(bsoncxx::builder::concatenate_doc{session->_get_impl().to_document()});
+        }
+
+        if (options.read_preference()) {
+            rp_ptr = options.read_preference()->_impl->read_preference_t;
+        }
+
+        libbson::scoped_bson_t opts_bson(opts_doc.view());
+
+        return libmongoc::collection_aggregate(
+            _coll, static_cast<::mongoc_query_flags_t>(0), stages.bson(), opts_bson.bson(), rp_ptr);
+    }
+
+    bsoncxx::stdx::optional<bsoncxx::string::view_or_value> create_one(
+        const client_session* session, const search_index_model& model) {
+        bsoncxx::document::value result =
+            create_many(session, std::vector<search_index_model>{model});
+        bsoncxx::document::view result_view = result.view();
+
+        return bsoncxx::stdx::make_optional(
+            bsoncxx::string::view_or_value(result_view["indexesCreated"]
+                                               .get_array()
+                                               .value.begin()
+                                               ->get_document()
+                                               .value["name"]
+                                               .get_string()
+                                               .value));
```

Comment:
```suggestion
        const auto result = create_many(session, std::vector<search_index_model>{model});
        return bsoncxx::stdx::make_optional(bsoncxx::string::view_or_value(result["indexesCreated"]
                                                                               .get_array()
                                                                               .value.begin()
                                                                               ->get_document()
                                                                               .value["name"]
                                                                               .get_string()
                                                                               .value));
```
Intermediate `result_view` does not seem necessary.

This function appears to always return a string value, otherwise `.get_string()` or any other document access prior will throw an exception. The [spec](https://github.com/mongodb/specifications/blob/9d7ceed8c37bafb5769787c26c8f8c79670f94f8/source/index-management/index-management.rst#index-view-api) also suggests the result is not expected to be optional. Consider changing return type from `bsoncxx::stdx::optional<bsoncxx::string::view_or_value>` to `bsoncxx::string::value`.

Reply:
Whoops. Intended to mean plain `std::string` (the `value` case of `view_or_value`).

Reply:
Right, sometimes I take things too literally :sweat_smile: